File: app/src/main/java/com/verbatoria/presentation/late_send/presenter/LateSendPresenter.java
package com.verbatoria.presentation.late_send.presenter;

import android.os.Bundle;
import android.support.annotation.NonNull;

import com.verbatoria.business.late_send.ILateSendInteractor;
import com.verbatoria.business.late_send.models.LateReportModel;
import com.verbatoria.business.session.ISessionInteractor;
import com.verbatoria.infrastructure.BasePresenter;
import com.verbatoria.presentation.late_send.view.ILateSendView;

import java.util.List;

/**
 * Реализация презентера для экрана поздней отправки
 *
 * @author nikitaremnev
 */
public class LateSendPresenter extends BasePresenter implements ILateSendPresenter {

    private static final String TAG = LateSendPresenter.class.getSimpleName();

    private ILateSendInteractor mLateSendInteractor;
    private ISessionInteractor mSessionInteractor;
    private ILateSendView mLateSendView;

    private List<LateReportModel> mLateReportModels;

    private int mSelectedPosition;

    public LateSendPresenter(ILateSendInteractor lateSendInteractor,
                             ISessionInteractor sessionInteractor) {
        mLateSendInteractor = lateSendInteractor;
        mSessionInteractor = sessionInteractor;
    }

    @Override
    protected void onStart() {
        super.onStart();
        mLateSendInteractor.getLateReports()
                .doOnSubscribe(() -> mLateSendView.showProgress())
                .doOnUnsubscribe(() -> mLateSendView.hideProgress())
                .subscribe(this::handleLateReportsReceived, this::handleLateReportsError);
    }

    @Override
    public void saveState(Bundle outState) {

    }

    @Override
    public void restoreState(Bundle savedInstanceState) {

    }

    @Override
    public void bindView(@NonNull ILateSendView lateSendView) {
        mLateSendView = lateSendView;
    }

    @Override
    public void unbindView() {
        mLateSendView = null;
    }

    @Override
    public void sendReport(int position) {
        mSelectedPosition = position;
        mLateSendView.showProgress();
        mSessionInteractor.submitResults(mLateReportModels.get(mSelectedPosition).getReportFileName())
                .subscribe(this::handleResultsSubmitted, this::handleSendError);
    }

    private void handleLateReportsReceived(List<LateReportModel> lateReportModelList) {
        mLateReportModels = lateReportModelList;
        if (mLateReportModels.isEmpty()) {
            mLateSendView.showNoReportsToSend();
        } else {
            mLateSendView.showLateReports(mLateReportModels);
        }
    }

    private void handleSendError(Throwable throwable) {
        mLateSendView.hideProgress();
        mLateSendView.showError(throwable.getLocalizedMessage());
    }

    private void handleLateReportsError(Throwable throwable) {
        throwable.printStackTrace();
        mLateSendView.hideProgress();
        mLateSendView.showNoReportsToSend();
        mLateSendView.showError(throwable.getLocalizedMessage());
    }

    private void handleResultsSubmitted() {
        mSessionInteractor.finishSession(mLateReportModels.get(mSelectedPosition).getSessionId())
                .subscribe(this::handleSessionFinished, this::handleSendError);
    }

    private void handleSessionFinished() {
        mLateSendInteractor.cleanUp(mLateReportModels.get(mSelectedPosition))
                .subscribe(this::cleanUpFinished, this::handleSendError);
    }

    private void cleanUpFinished() {
        mLateSendView.hideProgress();
        mLateReportModels.remove(mSelectedPosition);
        mLateSendView.notifyItemSent(mSelectedPosition);
        if (mLateReportModels.size() == 0) {
            mLateSendView.showNoReportsToSend();
        }
    }

}

File: app/src/main/java/com/verbatoria/presentation/late_send/LateSendPresenter.kt
package com.verbatoria.presentation.late_send

import com.verbatoria.business.late_send.LateSendInteractor
import com.verbatoria.business.late_send.models.LateReportModel
import com.verbatoria.business.session.SessionInteractor
import com.verbatoria.presentation.base.BasePresenter
import com.verbatoria.presentation.late_send.item.LateReportViewHolder

/**
 * @author n.remnev
 */

private const val NO_POSITION = -1

class LateSendPresenter(
    private var lateSendInteractor: LateSendInteractor,
    private var sessionInteractor: SessionInteractor
) : BasePresenter<LateSendView>(), LateReportViewHolder.Callback {

    private val lateReportsList: MutableList<LateReportModel> = mutableListOf()

    private var selectedPosition = NO_POSITION

    private var isLateReportsLoaded = false

    init {
        loadLateReports()
    }

    override fun onAttachView(view: LateSendView) {
        super.onAttachView(view)
        if (isLateReportsLoaded) {
            updateLateReportsList()
            view.hideProgress()
        } else {
            view.showProgress()
        }
    }

    //region LateReportViewHolder.Callback

    override fun onLateReportClicked(position: Int) {
        //empty
    }

    //endregion

    private fun loadLateReports() {
        view?.showProgress()
        lateSendInteractor
            .lateReports
            .subscribe(
                { lateReports ->
                    lateReportsList.addAll(lateReports)
                    updateLateReportsList()
                    view?.hideProgress()
                },
                { error ->
                    view?.showLateReportsIsEmpty()
                    view?.hideProgress()
                    view?.showErrorSnackbar(error.localizedMessage)
                }
            )
//            .let(::addDisposable)
    }

    private fun submitResults() {
        sessionInteractor
            .submitResults(lateReportsList[selectedPosition].reportFileName)
            .subscribe(
                {
                    finishSession()
                },
                { error ->
                    view?.hideProgress()
                    view?.showErrorSnackbar(error.localizedMessage)
                }
            )
        //            .let(::addDisposable)

    }

    private fun finishSession() {
        sessionInteractor
            .finishSession(lateReportsList[selectedPosition].sessionId)
            .subscribe(
                {
                    cleanUp()
                },
                { error ->
                    view?.hideProgress()
                    view?.showErrorSnackbar(error.localizedMessage)
                }
            )
        //            .let(::addDisposable)

    }

    private fun cleanUp() {
        sessionInteractor
            .cleanUp()
            .subscribe(
                {
                    lateReportsList.removeAt(selectedPosition)
                    view?.updateLateReportsList(lateReportsList)
                    if (lateReportsList.isEmpty()) {
                        view?.showLateReportsIsEmpty()
                    }
                    view?.hideProgress()
                },
                { error ->
                    view?.hideProgress()
                    view?.showErrorSnackbar(error.localizedMessage)
                }
            )
        //            .let(::addDisposable)
    }

    private fun updateLateReportsList() {
        if (lateReportsList.isEmpty()) {
            view?.showLateReportsIsEmpty()
        } else {
            view?.updateLateReportsList(lateReportsList)
        }
    }

}
